README.md
# Create CloudFront Custom Policy signed URLs in Laravel 6.0+

Easy to use Laravel 6+ | 7+ | 8+ wrapper which allows to sign URLs to access Private Content through CloudFront CDN

This package can create custom policy signed URLs for CloudFront which expires after a given time. The default time is set to 86400 seconds or 24 hours. This package does not depends on AWS SDK package, but implements the example from (https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/CreateURL_PHP.html#sample-custom-policy)

The benefit of this approach is that you can use the same signature for multiple files and hence the processing is much much faster when you are loading multiple signed images/videos or other files on a single page.

This is how you can create signed URL that's valid for 1 day or 24 hour:

## Usage

### Signing URLs

```php
// Add Namespace in Controller
use EpiphanyInfotech\CustomPolicyUrl\CustomPolicyUrl as AWSURL;

/*
 * Instantiate the class and pass the array with values matching 
 * the path of the media files in the AWS S3 Bubket
 */

        $key_id = config('aws-custom-policy-url.KEY-ID');
		
    	$aws_private_key = config('aws-custom-policy-url.AWS-PRIVATE-KEY');
    	
    	$path_arr = array('sample_5mb.mp4','sample_3mb.mp4'); // get the sample files either from your own DB or changing the storage config to s3 bucket and fetching the list of files. Ref.: (https://laravel.com/docs/8.x/filesystem#retrieving-files)

     	$aws_obj = New AWSURL($key_id, $aws_private_key);
    	$get_signed_urls = $aws_obj->getSignedUrls($path_arr);

```

The output is compliant with [CloudFront specifications](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-custom-policy.html)

## Installation

The package can be installed via Composer:

```
composer require epiphanyinfotech/laravel-aws-custom-policy-url-signer:^1.0
```

## Configuration

The configuration file can optionally be published via:

```
php artisan vendor:publish --provider="EpiphanyInfotech\CustomPolicyUrl\CustomPolicyUrlServiceProvider"
```

This is the content of the file:

```php
return [
    /*
     * The cloudfront public keys id. Identifies the CloudFront key pair associated to the trusted signer which validates signed URLs.
     * To create a key, refer: https://console.aws.amazon.com/cloudfront/v3/home?region=us-east-2#/publickey
     * Make sure to replace region=us-east-2 with your region
     */
    'KEY-ID' => env("AWS_KEY_ID", NULL),

    /*
     * The private key used to sign all URLs. Make sure it's secured and not accessible to public. Enter absolute path in the .env file
     */
    "AWS-PRIVATE-KEY" => env("AWS_PRIVATE_KEY_FILEPATH", NULL),

    /*
     * The IP address of the client making the GET request. For further info, visit: (https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-custom-policy.html) and search for - "IpAddress (Optional)"
     */
    "CLIENT_IP" => env("AWS_CLIENT_IP", NULL),,

    /*
     * Your CloudFront domain URL. To check your cloudfront domain, visit: (https://console.aws.amazon.com/cloudfront/v3/home?region=us-east-2#/distributions)
	 * Make sure to replace region=us-east-2 with your region.
	 * The "Domain Name" is your URL. If you have configured cloudfront to use the custom domain, then enter that.
     */
    "AWS_URL" => env("AWS_URL", NULL),

];
```


## Security

If you discover any security related issues, please email info@epiphanyinfotech.com or use the issue tracker.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `custom-policy-url` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


src/CustomPolicyUrlServiceProvider.php
<?php

namespace EpiphanyInfotech\CustomPolicyUrl;

use Illuminate\Support\ServiceProvider;

class CustomPolicyUrlServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        /*
         * Optional methods to load your package assets
         */
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'custom-policy-url');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'custom-policy-url');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('aws-custom-policy-url.php'),
            ], 'config');

            // Publishing the views.
            /*$this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/custom-policy-url'),
            ], 'views');*/

            // Publishing assets.
            /*$this->publishes([
                __DIR__.'/../resources/assets' => public_path('vendor/custom-policy-url'),
            ], 'assets');*/

            // Publishing the translation files.
            /*$this->publishes([
                __DIR__.'/../resources/lang' => resource_path('lang/vendor/custom-policy-url'),
            ], 'lang');*/

            // Registering package commands.
            // $this->commands([]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'custom-policy-url');

        // Register the main class to use with the facade
        $this->app->singleton('custom-policy-url', function () {
            return new CustomPolicyUrl;
        });
    }
}


src/CustomPolicyUrl.php
<?php

namespace EpiphanyInfotech\CustomPolicyUrl;

use Storage;

class CustomPolicyUrl
{
    // Build your next great package.

	public $key_id;
	public $aws_private_key;
	public $client_ip;
	public $aws_url;

    public function __construct($KEY_ID = NULL, $AWS_PRIVATE_KEY=NULL, $client_ip = NULL){
    	$this->key_id = $KEY_ID ?? config('aws-custom-policy-url.KEY_ID');
    	$client_ip = $client_ip ?? config('aws-custom-policy-url.CLIENT_IP');
    	$client_ip = $client_ip ?? $_SERVER['REMOTE_ADDR'];

    	$this->aws_private_key = $AWS_PRIVATE_KEY ??  config('aws-custom-policy-url.AWS_PRIVATE_KEY');
    	$this->client_ip = $client_ip;
    	$this->aws_url = config('aws-custom-policy-url.AWS_URL');
	}

	public function rsa_sha1_sign($policy, $private_key_filename) {
	    $signature = "";

	    // load the private key
	    $fp = fopen($private_key_filename, "r");
	    $priv_key = fread($fp, 8192);
	    fclose($fp);
	    $pkeyid = openssl_get_privatekey($priv_key);

	    // compute signature
	    openssl_sign($policy, $signature, $pkeyid);

	    // free the key from memory
	    openssl_free_key($pkeyid);

	    return $signature;
	}

	public function url_safe_base64_encode($value) {
	    $encoded = base64_encode($value);
	    // replace unsafe characters +, = and / with the safe characters -, _ and ~
	    return str_replace(
	        array('+', '=', '/'),
	        array('-', '_', '~'),
	        $encoded);
	}

	public function create_stream_name($stream, $policy, $signature, $key_pair_id, $expires) {
	    $result = $stream;
	    // if the stream already contains query parameters, attach the new query parameters to the end
	    // otherwise, add the query parameters
	    $path = '';
	    $separator = strpos($stream, '?') == FALSE ? '?' : '&';
	    // the presence of an expires time means we're using a canned policy
	    if($expires) {
	        $result .= $path . $separator . "Expires=" . $expires . "&Signature=" . $signature . "&Key-Pair-Id=" . $key_pair_id;
	    }
	    // not using a canned policy, include the policy itself in the stream name
	    else {
	        $result .= $path . $separator . "Policy=" . $policy . "&Signature=" . $signature . "&Key-Pair-Id=" . $key_pair_id;
	    }

	    // new lines would break us, so remove them
	    return str_replace('\n', '', $result);
	}

	public function encode_query_params($stream_name) {
	    // Adobe Flash Player has trouble with query parameters being passed into it,
	    // so replace the bad characters with their URL-encoded forms
	    return str_replace(
	        array('?', '=', '&'),
	        array('%3F', '%3D', '%26'),
	        $stream_name);
	}

	public function get_custom_policy_stream_name($video_path, $private_key_filename, $key_pair_id, $policy) {
	    // the policy contains characters that cannot be part of a URL, so we base64 encode it
	    $encoded_policy = $this->url_safe_base64_encode($policy);
	    // sign the original policy, not the encoded version
	    $signature = $this->rsa_sha1_sign($policy, $private_key_filename);
	    // make the signature safe to be included in a URL
	    $encoded_signature = $this->url_safe_base64_encode($signature);

	    // combine the above into a stream name
	    $stream_name = $this->create_stream_name($video_path, $encoded_policy, $encoded_signature, $key_pair_id, null);
	    // URL-encode the query string characters to support Flash Player
	    return $this->encode_query_params($stream_name);
	}

	public function customUrl($expiry_time){

		$private_key_filename = $this->aws_private_key;
		$key_pair_id = $this->key_id;
		$video_path = '';

		$expires = time() + $expiry_time; // Default 10 min from now
		
		$client_ip = $this->client_ip;
		$policy =
		'{'.
		    '"Statement":['.
		        '{'.
		            '"Resource":"'. $video_path . '",'.
		            '"Condition":{'.
		                '"IpAddress":{"AWS:SourceIp":"' . $client_ip . '/32"},'.
		                '"DateLessThan":{"AWS:EpochTime":' . $expires . '}'.
		            '}'.
		        '}'.
		    ']' .
		  '}';
		$custom_policy_stream_name = $this->get_custom_policy_stream_name($video_path, $private_key_filename, $key_pair_id, $policy);

		return urldecode($custom_policy_stream_name);
	}


	/*
	 * Function to return the AWS cloudfront encryptrd URL 
	 * Input: (array) | (string) $path_arr, (int) $expiry_time [in seconds] [Default 86400 second or 24 hours]
	 * Output: (array) Policy Appended URL
	 * - By Manpreet Kaur
	 */
	public function getSignedUrls($path_arr, $expiry_time = 86400){
		
		$aws_url = $this->aws_url;

		$url = $this->customUrl($expiry_time);

		$url_arr =  array();
		
		if(is_array($path_arr)){
			
			foreach($path_arr as $path){
	  		
		     	$url_arr[] = $aws_url.'/'. $path . $url;

			}	
		}else{
			
			$url_arr = $aws_url.'/'. $path_arr . $url;
		}
		
		return $url_arr;

	}

	public function test(){
		echo "You have reached the AWS package.";
	}


}
